### ISSUE TITLE
Workspace, repository, and label terms are still confusing, esp. w.r.t. repostiory_name

### ISSUE THREAD

**Issue by EricCousineau-TRI** [bazelbuild/bazel#16210]
### Description of the feature request:

It is a bit confusing two grok the two differences:
- `Label.workspace_name` does not have `@` in its name. However, `native.repository_name()` does have `@` in its name. Is this nuance documented? (both are stated as "names")
- I see the components that form a Label. However, it's a bit hard to denote the difference between "just" a package name (w.r.t. the 3 components) vs. a "qualified" package name (e.g. label of form `{qualified_package}:{name}`). Is there guidance on this?

The following docs are uber helpful:
- https://bazel.build/concepts/labels
- https://bazel.build/concepts/build-ref
- https://bazel.build/rules/lib/native#repository_name
- https://bazel.build/rules/lib/Label#workspace_name

But it'd still be nice to have something that denotes the minor difference, as well indicates some of term like "qualified package"  to denote `@workspace//package`.

### What underlying problem are you trying to solve with this feature?

Making terminology and usages clear and consistent eventually, and sifting through workarounds to #16209 (if that issue is actually valid).

### Which operating system are you running Bazel on?

Ubuntu 20.04

### What is the output of `bazel info release`?

release 5.1.0

### What's the output of `git remote get-url origin; git rev-parse master; git rev-parse HEAD` ?

```text
git@github.com:EricCousineau-TRI/repro.git
3c8f9afe4b0498b2920c262aebc5a756f75714bc
3c8f9afe4b0498b2920c262aebc5a756f75714bc
```

### Have you found anything relevant by searching the web?

see above

### Any other information, logs, or outputs that you want to share?

If at all helpful, this is some dumb heuristics from the following code (uncommenting certain pieces of code)
https://github.com/EricCousineau-TRI/repro/blob/3c8f9afe4b0498b2920c262aebc5a756f75714bc/bazel/label_stuff/defs.bzl#L22-L95
```
Label decomposition
//:lib:
  workspace:
  package:
  name: lib
//pkg:pkg:
  workspace:
  package: pkg
  name: pkg
//pkg:lib:
  workspace:
  package: pkg
  name: lib
@repo//:lib:
  workspace: repo
  package:
  name: lib
@repo//pkg:pkg:
  workspace: repo
  package: pkg
  name: pkg
@repo//pkg:lib:
  workspace: repo
  package: pkg
  name: lib

native.x() output:
In //pkg
  repository_name: @
  package_name: pkg
In //
  repository_name: @
  package_name:
In @repo//pkg
  repository_name: @repo
  package_name: pkg
In @repo
  repository_name: @repo
  package_name:
```

**Comment by Wyverald** [comment]
Thanks for filing this. I'll definitely take some time to tackle this for 6.0.0, together with https://github.com/bazelbuild/bazel/issues/15821

**Comment by EricCousineau-TRI** [comment]
Sweet, thanks!

If it helps at all, here was motivating application (or attempt haha):
EDIT: Usage https://github.com/EricCousineau-TRI/repro/tree/cf66d0d27ca9faad922d38f343bd693a43537483/drake_stuff/drake_downstream_pybind

**Comment by Wyverald** [comment]
Taking this off the milestone for now since it doesn't block branch cut. Will have to finish this before 6.0 launch though.

**Comment by Wyverald** [comment]
@bazel-io fork 7.1.0

**Comment by fmeum** [comment]
@Wyverald Could we also backport this to 6.5.0? Otherwise we will have to wait for one more year before we can disable the old API and/or allow rulesets to drop the feature detection required to work with both.

**Comment by Wyverald** [comment]
@bazel-io fork 6.5.0

**Comment by Wyverald** [comment]
Note -- there's another commit incoming that'll rename the new `Label.local_target_label` to `Label.same_package_label`. We should remember to cherry-pick that one into both 7.1.0 and 6.5.0 too.

**Comment by iancha1992** [comment]
> Note -- there's another commit incoming that'll rename the new `Label.local_target_label` to `Label.same_package_label`. We should remember to cherry-pick that one into both 7.1.0 and 6.5.0 too.

@Wyverald I just wanted to follow up on this. Do you want us to cherry-pick https://github.com/bazelbuild/bazel/commit/6cb2d860b025b0cfdbadb4943c8eabaff3497696 now to 7.1.0 and 6.5.0? Or are you pushing more changes?

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 6.5.0 RC1](https://releases.bazel.build/6.5.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
